File: 481 What Goes Up/solution.c
#include <stdio.h>
#include <stdlib.h>

typedef struct {
    int lastIndex, lis;
} DP;

int binarySearch(DP *array, int size, int key) {
    int left = 0, right = size - 1;
    while(left <= right) {
        int mid = left + (right - left) / 2;
        if(array[mid].lis < key)
            left = mid + 1;
        else
            right = mid - 1;
    }
    return left;
}

int main() {
	int *sequence = NULL, size = 0, num;
	while(scanf("%d", &num) == 1) {
	    sequence = (int*)realloc(sequence, (size + 1) * sizeof(int));
	    sequence[size++] = num;
	}
	DP *dp = NULL;
	int predecessor[size], lisSize = 0;
	for(int i = 0; i < size; ++i) {
	    int pos = binarySearch(dp, lisSize, sequence[i]);
	    if(pos == lisSize)
	        dp = (DP*)realloc(dp, ++lisSize * sizeof(DP));
	    dp[pos].lastIndex = i;
	    dp[pos].lis = sequence[i];
	    predecessor[i] = pos > 0 ? dp[pos - 1].lastIndex : -1;
	}
	int ans[lisSize], i = dp[lisSize - 1].lastIndex;
	for(int j = lisSize - 1; j >= 0; --j) {
	    ans[j] = sequence[i];
	    i = predecessor[i];
	}
	printf("%d\n-\n", lisSize);
	for(i = 0; i < lisSize; ++i)
	    printf("%d\n", ans[i]);
	free(sequence);
	free(dp);
	return 0;
}

File: 481 What Goes Up/solution.java
import java.io.*;
import java.util.*;

public class Main {
	public static void main(String[] args) throws IOException {
        ArrayList<Integer> sequence = new ArrayList<>();
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		StreamTokenizer st = new StreamTokenizer(br);
		while(st.nextToken() == StreamTokenizer.TT_NUMBER)
		    sequence.add((int)st.nval);
		ArrayList<DP> dp = new ArrayList<>();
		int[] predecessor = new int[sequence.size()];
		for(int i = 0; i < sequence.size(); ++i) {
		    int pos = binarySearch(dp, sequence.get(i));
		    if(pos == dp.size())
		        dp.add(new DP(i, sequence.get(i)));
		    else 
		        dp.get(pos).lis = sequence.get(dp.get(pos).lastIndex = i);
		    predecessor[i] = pos > 0 ? dp.get(pos - 1).lastIndex : -1;
		}
		int[] ans = new int[dp.size()];
		int i = dp.get(dp.size() - 1).lastIndex;
		for(int j = dp.size() - 1; j >= 0; --j) {
		    ans[j] = sequence.get(i);
		    i = predecessor[i];
		}
		StringBuilder output = new StringBuilder().append(lisSize).append("\n-\n");
		for(Integer j : ans)
		    output.append(j).append('\n');
        System.out.print(output);
	}
	
	static int binarySearch(ArrayList<DP> arrayList, int key) {
        int left = 0, right = arrayList.size() - 1;
        while(left <= right) {
            int mid = left + (right - left >> 1);
            if(arrayList.get(mid).lis < key)
                left = mid + 1;
            else
                right = mid - 1;
        }
        return left;
    }
}

class DP {
    int lastIndex, lis;
    DP(int lastIndex, int lis) {
        this.lastIndex = lastIndex;
        this.lis = lis;
    }
}
